curvefs/src/mds/topology/topology_manager.cpp

```diff
@@ -693,55 +693,74 @@ void TopologyManager::CreatePartitions(const CreatePartitionRequest *request,
-            LOG(ERROR) << "Get metaserver failed.";
-            return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
+
+        Pool pool;
+        if (!topology_->GetPool(poolId, &pool)) {
+            LOG(ERROR) << "Get pool failed.";
+            return TopoStatusCode::TOPO_POOL_NOT_FOUND;
         }
-        metaServerIds.emplace(metaServerId);
-    }
 
-    if (metaServerAddrs.size() != replicaNum) {
-        LOG(ERROR) << "Have invalid metaserver number when create copyset."
-                   << " metaserver number = " << metaServerAddrs.size()
-                   << ", replicanum = " << replicaNum;
-    }
+        replicaNum = pool.GetReplicaNum();
+        int needZoneNum = replicaNum;
+        std::set<ZoneIdType> zones;
+        std::set<ZoneIdType> unavailableZones;
+        do {
+            ret = topology_->ChooseZonesInPool(poolId, &zones, unavailableZones,
+                                            needZoneNum);
+            if (TopoStatusCode::TOPO_OK != ret) {
+                LOG(ERROR) << "Select Zone failed when create partition"
+                           << "poolId = " << poolId
+                           << ", erro msg = " << TopoStatusCode_Name(ret);
+                unavailablePools.emplace(poolId);
+                metaServerIds.clear();
+                break;
+            }
+
+            for (auto zoneId : zones) {
+                MetaServerIdType metaServerId;
+                ret = topology_->ChooseSingleMetaServerInZone(zoneId,
+                                                        &metaServerId);
+                if (TopoStatusCode::TOPO_OK != ret) {
+                    LOG(WARNING) << "Select metaServer failed when create"
+                                 << " partition zoneId = " << zoneId
+                                 << ", erro msg = " << TopoStatusCode_Name(ret);
+                    unavailableZones.emplace(zoneId);
+                    continue;
+                }
+
+                MetaServer metaServer;
+                if (topology_->GetMetaServer(metaServerId, &metaServer)) {
+                    metaServerAddrs.emplace(
+                        metaServer.GetInternalHostIp() + ":" +
+                        std::to_string(metaServer.GetInternalPort()));
+                } else {
+                    LOG(ERROR) << "Get metaserver failed.";
+                    return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
+                }
+                metaServerIds.emplace(metaServerId);
+            }
+            needZoneNum = unavailableZones.size();
+
+            if (metaServerAddrs.size() != replicaNum) {
+                LOG(WARNING) << "Have invalid metaserver number when create"
+                             << " copyset. metaserver number = "
+                             << metaServerAddrs.size()
+                             << ", replicanum = " << replicaNum;
+            }
+        } while (metaServerAddrs.size() < replicaNum);
+    } while (metaServerAddrs.size() < replicaNum);
```

Comment:
copyset can cross pool ?

Reply:
> copyset can cross pool ?

The copyset wiil create in the same pool, if the current pool have no available zones and before choose another pool, the metaserve have been selected will be cleared in line727.